cpp/src/io/text/multibyte_split.cu

```diff
@@ -275,180 +517,126 @@ int64_t multibyte_split_scan_full_source(cudf::io::text::data_chunk_source const
-                                                    string_offsets.end() - 1,
-                                                    byte_range.offset());
-
-  // This finds the first field beginning after the byte range.
-  // We shift it by 1 to also copy this last offset
-  auto relevant_offsets_end = 1 + thrust::lower_bound(rmm::exec_policy(stream),
-                                                      string_offsets.begin(),
-                                                      string_offsets.end() - 1,
-                                                      byte_range.offset() + byte_range.size());
-
-  // The above logic works if there are no duplicate string_offsets entries.
-  // The only way we can get duplicates is if the input ends with a delimiter.
-  // relevant_offsets_end should then point to the last entry, not the second-to-last, which can
-  // happen if byte_range.offset() + byte_range.size() matches the input size exactly.
-  // Without this adjustment, string_offsets_out would be missing the last element.
-  bool last_field_empty = string_offsets.size() >= 2 &&
-                          string_offsets.element(string_offsets.size() - 2, stream) == bytes_total;
-  bool byte_range_exact_end = byte_range.offset() + byte_range.size() == bytes_total;
-  if (last_field_empty && byte_range_exact_end) { ++relevant_offsets_end; }
-
-  auto string_offsets_out_size = relevant_offsets_end - relevant_offsets_begin;
-
-  auto string_offsets_out = rmm::device_uvector<int32_t>(string_offsets_out_size, stream, mr);
-
-  auto relevant_offset_first =
-    string_offsets.element(relevant_offsets_begin - string_offsets.begin(), stream);
-  auto relevant_offset_last =
-    string_offsets.element(relevant_offsets_end - string_offsets.begin() - 1, stream);
-
-  auto string_chars_size = relevant_offset_last - relevant_offset_first;
-  auto string_chars      = rmm::device_uvector<char>(string_chars_size, stream, mr);
-
-  // copy relevant offsets and adjust them to be zero-based.
-  thrust::transform(rmm::exec_policy(stream),
-                    relevant_offsets_begin,
-                    relevant_offsets_end,
-                    string_offsets_out.begin(),
-                    [relevant_offset_first] __device__(int64_t offset) {
-                      return static_cast<int32_t>(offset - relevant_offset_first);
+  auto chars          = char_storage.gather(stream, mr);
+  auto global_offsets = offset_storage.gather(stream, mr);
+
+  bool const insert_begin = *first_offset == 0;
+  bool const insert_end   = !last_offset.has_value() || last_offset == chunk_offset;
+  rmm::device_uvector<int32_t> offsets{
+    global_offsets.size() + insert_begin + insert_end, stream, mr};
+  if (insert_begin) { offsets.set_element_to_zero_async(0, stream); }
+  if (insert_end) { offsets.set_element(offsets.size() - 1, chunk_offset - *first_offset, stream); }
+  thrust::transform(rmm::exec_policy_nosync(stream),
```

Comment:
why nosync here?

Reply:
Why not? 😄 We don't write to or read from this range until we gather the data, so we don't need to wait for the kernels to finish before proceeding.